### ISSUE TITLE
Bug in MultiFactorScheduler implementation TF 

### ISSUE THREAD

**Issue by AnirudhDagar** [d2l-ai/d2l-en#1306]
In [11.11.3.2. Multi Factor Scheduler](http://preview.d2l.ai/d2l-en/master/chapter_optimization/lr-scheduler.html#multi-factor-scheduler) the `__call__` in the class defined will behave incorrectly when the user inputs an argument where size of list in `step` is more than 2.

```python
class MultiFactorScheduler:
  def __init__(self, step, factor, base_lr):
    self.step = step
    self.factor = factor
    self.base_lr = base_lr

  def __call__(self, epoch):
    if epoch in range(self.step[0], (self.step[1] + 1)):
      return self.base_lr * self.factor
    else:
      return self.base_lr

scheduler = MultiFactorScheduler(step=[15, 30], factor=0.5, base_lr=0.5)
d2l.plot(tf.range(num_epochs),
         [scheduler(t) for t in range(num_epochs)])
```

Eg if step = [10,15,30] the TF implementation is not correct for such case. 

PTAL @abhinavsp0730 @terrytangyuan 
cc @astonzhang

**Comment by StevenJokes** [comment]
@AnirudhDagar ,Would you like to help me to review my pytorch code before I PR to GAN.md ? It cost me whole night.
https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/my5gan.ipynb

**Comment by abhinavsp0730** [comment]
@Anirudh would you please tell me my whether my understanding is correct or not,
If the step is [15,30]
Then make the learning rate alph, b/w epochs 15 to 30.
If the step is [15,20,30],
Then make the learning rate, alpha only in the epochs {15}, {20} and {30} .
Please verify my understanding, If it is correct then the solution of this issue is just one If condition away.

**Comment by StevenJokes** [comment]
You @ the wrong guy.@abhinavsp0730 
@AnirudhDagar please

**Comment by AnirudhDagar** [comment]
@abhinavsp0730 [MultiFactorScheduler](https://mxnet.apache.org/versions/1.6/api/r/docs/api/mx.lr_scheduler.MultiFactorScheduler.html) or [MultiStepLR](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.MultiStepLR) decays the learning rate of each by gamma/factor once the number of epoch reaches one of the milestones/steps.

So you need to multiply the learning rate with the factor to update the learning rate once a particular step is reached.
Hence in [10,15,30] if we start with lr=0.5 and factor=0.1, then 
at epoch 10, lr=0.5*0.1=0.05
now this will continue to be the lr until next step is seen.
at epoch 15, lr=0.05*0.1=0.005
now this will continue to be the lr until next step is seen.
at epoch 30, lr=0.005*0.1=0.0005
and so on.
